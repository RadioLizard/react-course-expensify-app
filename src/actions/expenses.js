import { database} from '../firebase/firebase'

//ADD_EXPENSE
const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense
})

//REMOVE_EXPENSE
const removeExpense = ({id}={})=>({
    type: "REMOVE_EXPENSE",
    id
})

//START_REMOVE_EXPENSE
const startRemoveExpense=({id}={})=>{
    return(dispatch)=>{
        return database.ref(`expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense({id}))
        })   
    }
}
//EDIT_EXPENSE
const editExpense=(id, updates)=>({
    type: "EDIT_EXPENSE",
    id,
    updates
})

//SET_EXPENSES
const setExpenses=(expenses)=>({
    type: 'SET_EXPENSES',
    expenses
})

//START_SET_EXPENSES
const startSetExpenses=()=>{
    return(dispatch)=>{
         return database.ref('expenses')
        .once('value').then((snapshot)=>{
            const expenses=[]
            snapshot.forEach((childSnapshot)=>{
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}

//START_ADD_EXPENSE
const startAddExpense = (expenseData = {})=>{
    return (dispatch)=>{
        const  {
            description = '',
            note='',
            amount=0,
            createdAt=0
        } = expenseData
        const expense={description, note, amount, createdAt}
       return database.ref('expenses').push(expense).then((ref)=>{
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

export {removeExpense, editExpense, addExpense, startAddExpense, setExpenses, startSetExpenses, startRemoveExpense}