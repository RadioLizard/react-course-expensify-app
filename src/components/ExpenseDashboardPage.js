import React from 'react'
import ConnectedExpenseList from './ExpenseList'
import ConnectedExpenseListFilters from './ExpenseListFilters'

const ExpenseDashboardPage = () =>{
    return(
    <div>
        <p>This is my expense dashboard page</p>   
        <ConnectedExpenseList/>
        <ConnectedExpenseListFilters/>
    </div>
    )
}

export default ExpenseDashboardPage