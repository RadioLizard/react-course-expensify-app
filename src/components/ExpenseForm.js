import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'

class ExpenseForm extends React.Component{
    constructor(props){
    super(props)
    this.state={
        description: props.expense ? props.expense.description : '',
        amount: props.expense ? (props.expense.amount/ 100).toString() : 0,
        note: props.expense ? props.expense.note : '',
        createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
        focused:false,
        error:undefined
    }
}
    onDescriptionChange=(e)=>{
        const description = e.target.value
        this.setState(()=>({description}))
    }
    onNoteChange=(e)=>{
        const note=e.target.value
        this.setState(()=>({note}))
    }
    onAmountChange=(e)=>{
        const amount=e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            this.setState(()=>({amount}))
        }
    }
    onDateChange=(createdAt)=>{
        if(createdAt){
            this.setState(()=>({createdAt}))
        }
    }
    onFocusChange=({focused})=>{
        this.setState(()=>({focused}))
    }
    onSubmit=(e)=>{
        e.preventDefault()
        if(!this.state.description || !this.state.amount){
            this.setState(()=>({error: "Please add a valid description and amount."}))
        }
        else{
            this.setState(()=>({error:undefined}))
            this.props.onSubmit({
              description: this.state.description,
              amount: parseFloat(this.state.amount, 10)*100,
              createdAt: this.state.createdAt.valueOf(),
              note: this.state.note
            })
        }
    }
    render(){
        return(
            <div>
                {!!this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        value={this.state.description}
                        autoFocus
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="number"
                        placeholder="amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.focused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day)=>false}
                    />
                    <textarea
                        value={this.state.note}
                        placeholder="add a note for this expense"
                        onChange={this.onNoteChange}>
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm