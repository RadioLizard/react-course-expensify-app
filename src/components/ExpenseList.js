import React from 'react'
import {connect} from 'react-redux'
import Expense from './Expense'
import getVisibleExpenses from '../selectors/expenses'

const ExpenseList=(props)=>(
    <div>
        {
            props.expenses.length===0 ? (
                <p>No expenses to show</p>
            ):(
                props.expenses.map((expense, index)=>(
                <Expense
                    {...expense}
                    key={index}
                />))
            )
        }
    </div>
)

const mapStateToProps=(state)=>{
    return{
        expenses: getVisibleExpenses(state.expenses, state.filters),
    }
}
const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)

export default ConnectedExpenseList
export {ExpenseList}