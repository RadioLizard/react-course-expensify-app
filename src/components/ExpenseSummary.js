import React from 'react'
import {connect} from 'react-redux'
import  numeral from 'numeral'
import selectExpensesTotal from '../selectors/filters'
import getVisibleExpenses from '../selectors/expenses'

const ExpenseSummary = (props) =>(
    <div>
        <p>Viewing {props.expenseCount} 
            {props.expenseCount===1 ? 
            <span> expense</span> : 
            <span> expenses</span>}, 
            which {props.expenseCount===1 ?
            <span>totals </span>:
            <span>total </span>}
            {numeral(props.expenseTotal / 100).format('$0,0.00')}</p>
    </div>
)

const mapStateToProps=(state)=>{
    return{
        expenseTotal: selectExpensesTotal(getVisibleExpenses(state.expenses, state.filters)),
        expenseCount: getVisibleExpenses(state.expenses, state.filters).length
    }
}
const ConnectedExpenseSummary = connect(mapStateToProps)(ExpenseSummary)

export default ConnectedExpenseSummary
export{ExpenseSummary}