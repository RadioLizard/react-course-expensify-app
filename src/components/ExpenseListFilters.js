import React from 'react'
import {connect} from 'react-redux'
import {DateRangePicker} from 'react-dates'
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters'

class ExpenseListFilters extends React.Component{
    state={
        calandarFocused: null,
    }
    onDatesChange=({startDate, endDate})=>{
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange=(calandarFocused)=>{
        this.setState(()=>({calandarFocused}))
    }
    onTextChange=(e)=>{
        this.props.setTextFilter(e.target.value)
    }
    onSortChange=(e)=>{
        if(e.target.value==="date"){
            this.props.sortByDate()
        }
        else if(e.target.value==="amount"){
            this.props.sortByAmount()
        }
    }
    render(){
        return(
        <div> 
            <input 
                type="text" 
                value={this.props.filters.text} 
                onChange={this.onTextChange}
            />
            <p>Sort By:</p>
            <select value={this.props.filters.sortBy} onChange={this.onSortChange}> 
                <option value="date">date</option>
                <option value="amount">amount</option>
            </select>
            <DateRangePicker
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calandarFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={()=>false}
                showClearDates={true}
            />
        </div>
        )
    }
}

const mapStateToProps = (state) =>{
    return{
        filters:state.filters
    }
}

const mapDispatchToProps = (dispatch)=>({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

const ConnectedExpenseListFilters=connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)

export default ConnectedExpenseListFilters
export {ExpenseListFilters}