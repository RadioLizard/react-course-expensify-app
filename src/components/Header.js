import React from 'react'
import {Link} from 'react-router-dom'
import {connect} from 'react-redux'
import {startLogout} from '../actions/auth'

const Header =({startLogout})=>(
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/dashboard"> 
                    <h1>Expensify</h1>
                </Link>
                <button className="button button--link" onClick={startLogout}>Logout</button>
            </div>
        </div>
    </header>
)   

const mapDispatchToProps=(dispatch)=>{
    return{
        startLogout: ()=> dispatch(startLogout())
    }
}
const ConnectedHeader =  connect(undefined, mapDispatchToProps)(Header)

export {Header, ConnectedHeader}