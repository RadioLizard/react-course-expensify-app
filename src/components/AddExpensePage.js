import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { addExpense} from '../actions/expenses'

class AddExpensePage extends React.Component {
    addExpense=(expense)=>{
        this.props.addExpense(expense)
        this.props.history.push('/')
    }
    render(){
        return(
            <div>
                <h1>Add an Expense</h1>
                <ExpenseForm
                    onSubmit={this.addExpense}
                />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch)=>({
    addExpense: (expense)=> dispatch(addExpense(expense))
})
    
const ConnectedAddExpensePage=connect(undefined, mapDispatchToProps)(AddExpensePage)

export default ConnectedAddExpensePage
export{AddExpensePage}