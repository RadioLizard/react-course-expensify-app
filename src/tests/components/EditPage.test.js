
    //Test 3: Should remove expense (use spies, making sure they've been called)

    import React from 'react'
    import {shallow} from 'enzyme'
    import expenses from '../fixtures/expenses'
    import {EditPage} from '../../components/EditPage'
    

    let editExpense, removeExpense, wrapper, history

    beforeEach(()=>{
         editExpense=jest.fn()
         removeExpense=jest.fn()
         history={
             push: jest.fn()
         }
         wrapper=shallow(
             <EditPage 
                history={history} 
                editExpense={editExpense} 
                removeExpense={removeExpense} 
                expense={expenses[2]}
            />
        )
         
    })

    test('Should render edit expense page with snapshot',()=>{
        expect(wrapper).toMatchSnapshot()
    })

    test('Should handle edit expense correctly, running spy', ()=>{
       wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
       expect(history.push).toHaveBeenLastCalledWith('/')
       expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])
    })   

    test('Should handle remove expense with spy', ()=>{
        wrapper.find("button").simulate('click')
        expect(history.push).toHaveBeenLastCalledWith('/')
        expect(removeExpense).toHaveBeenLastCalledWith({id:expenses[2].id})
    })