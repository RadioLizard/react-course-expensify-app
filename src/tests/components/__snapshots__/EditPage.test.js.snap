// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render edit expense page with snapshot 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "credit card",
        "id": "3",
        "notes": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Delete Expense
  </button>
</div>
`;
